tfx/experimental/distributed_inference/graphdef_experiments/subgraph_partitioning/beam_pipeline.py

```diff
@@ -0,0 +1,246 @@
+# Copyright 2020 Google LLC. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Run batch inference in beam on partitioned graphs.
+
+After getting execution_specs from graph partitioning, we'd like to arrange,
+load, and execute the subgraphs in beam.
+
+Execution_specs:
+    <--closer to inputs                   closer to outputs-->
+        --------------------------------------------------
+        |          |          |          |               |
+        |  spec 1  |  spec 2  |  spec 3  | ......        |
+        |          |          |          |               |
+        --------------------------------------------------
+                 /              \
+                /                \
+               / ---------------- \
+                 |   Subgraph   |
+                 ----------------
+                 |    Inputs    |
+                 ----------------
+                 |   Outputs    |
+                 ----------------
+                 | is_remote_op |
+                 ----------------
+"""
+
+import copy
+import tensorflow as tf
+import apache_beam as beam
+
+
+@beam.ptransform_fn
+def ExecuteOneGraph(pcoll,  # pylint: disable=invalid-name
+                    op_to_execution_specs,
+                    op_to_remote_op_name_mapping,
+                    graph_name):
+  """Compute one graph, for example remote_op_b's graph.
+
+  Main assumption:
+    The parent graph has set up the placeholder inputs for the child graph.
+    (So we need to setup PColl for the main graph)
+
+  Args:
+    pcoll: input PCollection, each unit contains
+           {graph_name: {computed node: value}}
+    op_to_execution_specs: {graph_name: [spec1, spec2, ...]}
+    spec: {'subgraph': a graph_def,
+           'inputs': a set of node names,
+           'outputs': a set of node names,
+           'is_remote_op': a boolean}
+
+    # This information was stored in py_func's feed_dict,
+    # so we don't have it.
+    op_to_remote_op_name_mapping:
+      {graph_name:
+        {remote op name: {placeholder name inside subgraph: input name}}}
+
+    graph_name: which graph am I currently executing
+
+  Returns:
+    pcoll with the intermediate/end results of this graph.
+  """
+  execution_specs = _get_execution_specs(graph_name, op_to_execution_specs)
+
+  count = 0
+  for spec in execution_specs:
```

Comment:
Delete the `count` variable. If want to use the index, you can use `for i, spec in enumerate(execution_specs)`, but you should really do something more informative like use `graph_name` in the step name instead.

Reply:
Changed.